checkpoint/recovery.go

```diff
@@ -0,0 +1,383 @@
+			)
+		}
+		return dst, nil
+	}
+
+	if parsed.Scheme != "http" && parsed.Scheme != "https" {
+		return "", fmt.Errorf("uri scheme %s not supported", uri)
+	}
+	if bdir, err := backupRecovery(fs, RecoveryDir(dataDir)); err != nil {
+		return "", err
+	} else if bdir != "" {
+		logger.WithContext(ctx).With().Info("old recovery data backed up", log.String("dir", bdir))
+	}
+	if err = httpToLocalFile(ctx, parsed, fs, dst); err != nil {
+		return "", err
+	}
+	logger.WithContext(ctx).With().Info("checkpoint data persisted", log.String("file", dst))
+	return dst, nil
+}
+
+func Recover(
+	ctx context.Context,
+	logger log.Log,
+	fs afero.Fs,
+	cfg *RecoverConfig,
+	nodeID types.NodeID,
+	uri string,
+	restore types.LayerID,
+) error {
+	db, err := sql.Open("file:" + filepath.Join(cfg.DataDir, cfg.DbFile))
+	if err != nil {
+		return fmt.Errorf("open old database: %w", err)
+	}
+	newdb, err := RecoverWithDb(ctx, logger, db, fs, cfg, nodeID, uri, restore)
+	if err != nil {
+		return err
+	}
+	if err = db.Close(); err != nil {
+		return fmt.Errorf("close old db: %w", err)
+	}
+	if newdb != nil {
+		if err = newdb.Close(); err != nil {
+			return fmt.Errorf("close new db: %w", err)
+		}
+	}
+	return nil
+}
+
+func RecoverWithDb(
+	ctx context.Context,
+	logger log.Log,
+	db *sql.Database,
+	fs afero.Fs,
+	cfg *RecoverConfig,
+	nodeID types.NodeID,
+	uri string,
+	restore types.LayerID,
+) (*sql.Database, error) {
+	oldRestore, err := recovery.CheckpointInfo(db)
+	if err != nil {
+		return nil, fmt.Errorf("get last checkpoint: %w", err)
+	}
+	if oldRestore >= restore {
+		return nil, nil
+	}
+	logger.With().Info("Recover from uri", log.String("uri", uri))
+	cpfile, err := copyToLocalFile(ctx, logger, fs, cfg.DataDir, uri, restore)
+	if err != nil {
+		return nil, err
+	}
+	return recoverFromLocalFile(ctx, logger, db, fs, cfg, nodeID, cpfile, restore)
+}
+
+type recoverydata struct {
+	accounts []*types.Account
+	atxs     []*atxs.CheckpointAtx
+}
+
+type ownAtxData struct {
+	atx      *types.VerifiedActivationTx
+	preserve []*types.VerifiedActivationTx
+	proof    []byte
+}
+
+func preserveOwnData(
+	db *sql.Database,
+	cfg *RecoverConfig,
+	nodeID types.NodeID,
+	data *recoverydata,
+) (ownAtxData, error) {
+	var own ownAtxData
+	atxid, err := atxs.GetLastIDByNodeID(db, nodeID)
+	if err != nil && !errors.Is(err, sql.ErrNotFound) {
+		return own, fmt.Errorf("query own last atx id: %w", err)
+	}
+	if atxid == types.EmptyATXID {
+		return own, nil
+	}
+	all := map[types.ATXID]struct{}{}
+	for _, catx := range data.atxs {
+		all[catx.ID] = struct{}{}
+	}
+	if _, ok := all[atxid]; ok {
+		return own, nil
+	}
+
+	own.atx, err = atxs.Get(db, atxid)
+	if err != nil {
+		return own, fmt.Errorf("get own atx: %w", err)
+	}
+	own.preserve = append(own.preserve, own.atx)
+	deps := map[types.ATXID]struct{}{own.atx.PrevATXID: {}}
+	deps[own.atx.PositioningATX] = struct{}{}
+	if own.atx.CommitmentATX != nil {
+		deps[*own.atx.CommitmentATX] = struct{}{}
+	}
```

Comment:
it such atx is preserved and re-gossiped and commitment atx is not in the state, what do we do?

Reply:
do i understand it correctly, that you preserve whole chain and then nodes will have to download it? i think this is not an option

Reply:
i am somewhat confused about atxs that are kept for recovery. originally there was an idea that we keep a set of "golden" atxs, that are hardcoded when checkpoint is distributed. it meant that the only part that has to be kept around is post data.

later there was a re-gossip idea, but (re)-gossiping requires full body of the atx that was generated using old code. so e.g post proof, poet proof, other atx validation must be kept atleast for some time. is it definitely better?

Reply:
there are two kinds of ATXs:

A. checkpointed: 
  - these are the golden atxs you were referring to. 
  - the body is not preserved and these cannot be gossiped.
  - new miners have to get these atxs from the checkpoint file.
  - currently we take the latest 2 atxs from each miner to form the set of checkpointed atxs
 
B. re-gossiped own atxs:
  - these are full atxs that preserved at the time of recovering from checkpoint file when a miner finds that its own latest atx is not included in the checkpoint file.
  - these are re-gossiped via syncer syncing aggressively at the time of recovery
  - the whole chain is not preserved, only the directly referenced atxs, and those referenced ones should refer back to one in set A. 
 
the whole point of B is so that miners do not have to wait another epoch if their atxs missed the checkpoint. as you pointed out, atxs in group B will carry the old format, which may defeat the purpose of the checkpoint recovery. 

but the worst case scenario is similar, if they are no longer valid under the new code after checkpoint, then we drop them (without marking them malicious). if they are still valid, then we gain on UX.

in any case, we can drop B from this PR and initiate a separate discussion with research.

Reply:
i guess allowing them by default and disabling by config option seems reasonable